### DISCUSSION TITLE
How to write unit tests for scripts with `@onready` vars?

### DISCUSSION THREAD

**Discussion by bugbountyguy** [MikeSchulze/gdUnit4#330] [Q&A] [answered]
Hi there,

I've got a new script that I want to write unit tests for, but any time I write a test for any function that uses a variable from `@onready`, it throws an exception about it not being set.

For example, if I have:
```
@onready var _player_model: MeshInstance3D = $PlayerModel
```
and try to run a function that uses values or something from `_player_model`, I get an exception that `_player_model` is nil. 

How can I setup my unit tests to work with these? I couldn't find any examples of this or mentions of it in the documentation so thought I'd come here to ask :)

**Comment by MikeSchulze** [comment, marked as answer]
Hi @bugbountyguy i guess you have a scene with attached script you want to test.

The annotations are set when a node is added to the scenetree.
So you have two options.
a) load your scene and add it to the test-suite, with `add_child()` the `_ready` step is executed, and you should have a full instantiated model. 
**Please note that ALL dependencies of the scene are loaded here.**
If you only need a part of the dependencies, you should rather mock the relevant components. See option b
```
func test_xxx() -> void:
	var scene = load("res://my_scene.tscn").instantiate()
	add_child(scene)
	
	assert_that(scene._player_model).is_not_null()
```
b) Load your script you want to test and set the `_player_model` manually and use a mock to assign a default model.
```
class XXX extends Node:
	@onready var _player_model: MeshInstance3D = $PlayerModel
	
	func my_func_to_test() -> Material:
		return _player_model.get_active_material(0)

func test_xxx() -> void:
	var _class_to_test := auto_free(XXX.new())
	_class_to_test._player_model = mock(MeshInstance3D)
	
	assert_that(_class_to_test.my_func_to_test()).is_null()
```

**Reply by bugbountyguy** [reply]
As always, thank you so much for your quick responses and help. solution B was exactly what I needed! :)
